Add graphical layout system

MrMDavidson opened the pull request:
# What does this implement/fix?

Adds a simple graphical layout system. My experience making ESPhome devices with a physical UI is that there's a lot of trial and error to get things to layout nicely. I expect the API in this PR to evolve over time as I add different UI elements and trial it in different ways.

The intent of this is PR is to solicit feedback from the community.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3505

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

display:
#  ...
  pages:
    - id: layout_test_page
      lambda: |-
        it.clear();
        
        it.render_layout(0, 0, id(test_layout);

graphical_layout:
  id: test_layout
  layout:
    type: vertical_stack
    item_padding: 1
    items:
      - type: horizontal_stack
        item_padding: 0
        items: 
          - type: text_panel
            font: small_font
            text: 'Left to'
          - type: text_panel
            font: small_font
            text: ' right text'
      - type: text_panel
        font: big_font
        text: 'A 2nd row of text  '

```

Sample config produces: 
![image](https://github.com/esphome/esphome/assets/1366249/1c52d7fd-7f74-4a44-ad17-a5e2a26f0d63)

A more complicated example:

```yaml
graphical_layout:
  - id: climate_sample_layout
    layout:
      type: fixed_dimension_panel
      width: display
      height: display
      child:
        type: vertical_stack
        child_align: stretch_to_fit_width
        item_padding: 2
        items:
          # Row 1 - Header
          - type: horizontal_stack
            items:
              - type: text_panel
                font: big_font
                text: Climate Info
          # Row 2 Text + Graph
          - type: horizontal_stack
            items:
              # Row 2 - Column 1 - Text
              - type: text_run_panel
                max_width: 170
                border: 0
                margin: 4
                debug_outline_runs: false
                text_align: baseline_left
                runs:
                  - font: text_font
                    text: "The temperature in the "
                  - font: bold_font
                    text: !lambda |-
                      return "study";
                  - font: text_font
                    text: " is "
                  - font: bold_font
                    sensor: temperature
                    text_formatter: !lambda |-
                      return it + "°C";
                  - font: text_font
                    text: " and "
                  - font: bold_font
                    sensor: humidity
                    text_formatter: !lambda |-
                      return it + "%";
                  - font: text_font
                    text: " humidity. It is "
                  - font: bold_font
                    text_sensor: thermal_comfort
              # Row 2 - Column 2 - Graph
              - type: display_rendering_panel
                width: ${graph_width}
                height: ${graph_height}
                lambda: !lambda |-
                  it.graph(0, 0, id(climate_graph));
```
![image](https://github.com/esphome/esphome/assets/1366249/d3780ad4-2822-45f2-87f9-383353d5fee8)


## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

MrMDavidson committed:
Introduce a local coordinate drawing system to display and display_buffer

MrMDavidson committed:
Initial check in of a graphical layout system
WIP: But supports vertical and horizontal stacking of items and a simple text item

MrMDavidson committed:
Merge branch 'dev' of github.com:MrMDavidson/esphome into add-graphical-layout-system

# Conflicts:
#	CODEOWNERS
#	esphome/components/display/display.cpp
#	esphome/components/display/display.h
#	esphome/core/defines.h

MrMDavidson committed:
Lint/clang fixes

MrMDavidson committed:
More clang/ling fixes

MrMDavidson committed:
clang fixes

MrMDavidson committed:
Clang fixes

MrMDavidson committed:
... more clang fixes

MrMDavidson committed:
Some lint improvements

MrMDavidson committed:
Add first implementation of a DisplayRenderingPanel
Makes use of a standard display_writer_t style Lambda so users can use arbitrary APIs from the Display Rendering Engine within the layout engine

MrMDavidson committed:
Black, flake8, and clang fixes

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Make use of constant from const.py

MrMDavidson committed:
Move layout schemas into the layout items - this keeps them local to the config generation code that uses them

MrMDavidson committed:
Fix usage of CONF_TYPE

jesserockz commented:
My initial thought on this is that I would make this part of the `pages` config object, exclusive with `lambda`

MrMDavidson commented:
> My initial thought on this is that I would make this part of the `pages` config object, exclusive with `lambda`

So for dynamic scenarios the user would need a page per layout and switch between them as needed? Does your thought on having it be one or the other change with the inclusion of the `DisplayRenderingPanel` (final name to be determined!)
```yaml
  - type: display_rendering_panel
    height: 50
    width: 50
    lambda: |-
      it.filled_circle(10, 10, 10);
      it.filled_rectangle(30, 30, 10, 10);
      // Other arbitrary drawing
```

MrMDavidson committed:
Clang fixes

MrMDavidson committed:
Fixed const-ness of TAG in vertical_stack and horizontal_stack

MrMDavidson committed:
Add support for margin, padding, and border on the base LayoutItem
Margin is the space between items
Border is a an outline of the element
Padding is the space between the border and the actual drawing
This is supported by a measure/render pair that are implemented in LayoutItem which handle margin/border/padding this calls measure_internal/render_internal
Most implementations will simply override the _internal versions and make use of the existing drawing/calculations

MrMDavidson committed:
Add support for controlling child alignment in Horizontal and Vertical Stack
This allows the user to left/center/right/stretch items in the VerticalStack and top/center/bototm/stretch items in the HorizontalStack

MrMDavidson committed:
Black. flake8, pylint, and clang-format fixes

MrMDavidson committed:
Additional clang-format and ci-custom fixes

MrMDavidson committed:
Add support for text alignment to TextPanel

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Black, flake8, and clang-format fixes

MrMDavidson committed:
Fix some incorrectly templatable values as well as the implementation of items that should be templatable

MrMDavidson committed:
Provide a bounds to the DisplayRenderingPanel
This allows users to customise their drawing depending on the available space

MrMDavidson committed:
black / flake8 / pylint fixes

MrMDavidson committed:
Add the FixedDimensionPanel
This renders a single child (you'd likely want that child to be a Vertical or Horizontal Stack Panel) with constrained dimensions.

For example to achieve a fixed two column 50/50 display you could have a layout of;
Root
 - Horizontal Stack
   - Fixed Dimension (50% display width)
     - Vertical Stack
       - Text, Display, etc...
   - Fixed Dimension (50% display width)
     - Vertical Stack
       - More elements...

MrMDavidson committed:
Remove unused item_padding

MrMDavidson committed:
Add text_align_to_string
Used for debugging ESP_LOGCONFIG purposes

MrMDavidson committed:
Add setup_complete hook
Called at the end of setup() of the underlying Component. Last chance to do pre-run-time initialisation in LayoutItems

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Add [horizontal|vertical]_child_align_to_string methods
Used as part of ESP_LOGCONFIG() to aide in human readability

MrMDavidson committed:
Ensure child items have their measure_item() called so they can perform layout, if necessary

MrMDavidson committed:
Switch to using display's definition of COLOR_ON/COLOR_OFF + improve ESP_LOGCONFIG() output

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Add text_run_panel
Unlike text_panel (to be renamed) the text_run_panel supports multiple runs of different fonts (and thus sizes, weights, and colours) and will lay them out to the best of its ability - respecting desired alignment - as it can.

MrMDavidson committed:
Apply flake8. pylint, ci-custom, and clang-format fixes

nielsnl68 commented:
Hello @MrMDavidson last year i worked on an similar project named [SwitchPanel ](https://github.com/nielsnl68/esphome-components/tree/main/components/switchplate). Maybe you it has some stuff that is interesting for your project.
Also just for your information there is also some work to intergrade LVGL into esphome. 

Suc6 with your project.

MrMDavidson committed:
clang-format, ci-custom, and build fixes

MrMDavidson commented:
> Hello @MrMDavidson last year i worked on an similar project named [SwitchPanel ](https://github.com/nielsnl68/esphome-components/tree/main/components/switchplate). Maybe you it has some stuff that is interesting for your project. Also just for your information there is also some work to intergrade LVGL into esphome.
> 
> Suc6 with your project.

Interesting project - I remember you mentioning it when I was working on the graphical menu PR. Was there a reason you made it as an external component rather than part of esphome itself?

MrMDavidson committed:
Build fixes

MrMDavidson committed:
Add simple test layout

MrMDavidson committed:
Apply clang-tidy recommendations

MrMDavidson committed:
Test, clang-format, and ci-custom fixes

MrMDavidson committed:
Apply clang-tidy recommendation

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
clang-tidy fixes

MrMDavidson committed:
clang-format fixes

MrMDavidson committed:
Add dump_config_base_properties to LayoutItem - displays the basic properties of the item (Margin, Border, Padding)

MrMDavidson committed:
Add ability for FixedDimensionPanel to operate in one of three modes; display dimension, child dimension, or fixed.
- Display dimensions use the width/height of the display (minus margin, border, and padding) for the dimension
- Child dimensions fix the dimension to that of the child
- Fixed dimension uses a specified pixel value (or lambda returning pixel value)

MrMDavidson committed:
ci-custom and clang-format fixes

MrMDavidson committed:
More ci-custom + clang-format fixes

MrMDavidson committed:
Use local defintions of COLOR_ON / COLOR_OFF

MrMDavidson committed:
Add can_wrap_at_character - allows for overriding how lines are wrapped
This can be used if the default characters do not work for a user or to support languages other than English

MrMDavidson committed:
Make margin, borders, and padding individually settable (top vs left vs bottom vs right)
Can still take an int to set all the properties equally

MrMDavidson committed:
ci-custom and clang-format fixes

MrMDavidson committed:
Additional clang-format fixes

MrMDavidson committed:
Change to less-than-or-equal checks when laying out items
Previously this would cause the actual rendering of item to want different bounds as the longest line would reflow as it was equal to the bounds

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Add support for TextPanels grabbing their content from Sensors and TextSensors

MrMDavidson committed:
Add support for Text Run Panel grabbing run text from a Sensor or Text Sensor

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Apply build fixes, black, flake8, pylint, and ci-custom recommendation

MrMDavidson committed:
Clang-format fixes

MrMDavidson committed:
Apply clang-tidy fixes

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
clang-format fixes

MrMDavidson committed:
Fix clang-tidy issues

MrMDavidson committed:
Move graphical_layout test to test1.1 in an attempt to avoid "segment data does not fit" issues

MrMDavidson committed:
Include SPI for test1.1

MrMDavidson committed:
Include Roboto font in test1.1

MrMDavidson committed:
Add colour definitions to test1.1

MrMDavidson committed:
Change pin in test1.1 to avoid redefinition

MrMDavidson committed:
Change pin number

MrMDavidson committed:
Remove allow_other_uses

MrMDavidson committed:
Allow pin reuse in test1.1 to fix test errors

MrMDavidson committed:
Add allow_other_uses for GPIO05

MrMDavidson committed:
ci-custom fix

MrMDavidson committed:
Rework text splitting and layout algorithm to be easier to read and work on
Fixes an bug where text would sometimes overflow the x axis in some cases

MrMDavidson committed:
Apply ci-custom and clang-format fixes

MrMDavidson committed:
Apply clang-tidy fixes

MrMDavidson committed:
Clang-format fixes

MrMDavidson committed:
Clang-tidy fixes

MrMDavidson committed:
Autoload sensor and text_sensor

MrMDavidson committed:
Add support for time based Text Panel and Text Run Panels
Easy way to print times and dates in different formats

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Remove unused variable

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Ensure text does not render beyond the available bounds given to the TextRunPanel
Introduce draw_partial_lines option. If enabled (defaults to false) partial lines of text will be rendered to the screen

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
When fixed_dimension_panel uses the display width/height ensure it only uses the available remaining width/height based on where it's positioned

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Introduce calculated metadata about runs. Allows replacement (eg. \t to '    '), new line handling, and suppression of non-printable characters

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Clang tidy fixes

MrMDavidson committed:
Introduce ParagraphBreakTextRun. Forcefully causes new paragraphs

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson commented:
> My initial thought on this is that I would make this part of the `pages` config object, exclusive with `lambda`

Hey @jesserockz , any more thoughts on this?

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Switch to using shared_ptr for Text Runs being added to the TextRunPanel

MrMDavidson committed:
Merge branch 'add-graphical-layout-system' of github.com:MrMDavidson/esphome into add-graphical-layout-system

MrMDavidson committed:
Apply clang-tidy suggestions

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

GraffJosh commented:
I tried to use this PR, because I'm facing a very similar issue:
waveshare epd driver board
4.2" display WFT0420CZ15

When I install the yaml I get this:
![image](https://github.com/esphome/esphome/assets/15370055/2d0adbb4-9aaa-4ec4-9943-060f6187de82)
![image](https://github.com/esphome/esphome/assets/15370055/335ebc31-3ca7-4610-a4ef-4cc7641e3b58)

Thanks for your work thusfar, hoping there's a solution out there.

jesserockz committed:
Merge branch 'dev' into pr/MrMDavidson/5949

jesserockz committed:
Organise imports

jesserockz committed:
Merge branch 'dev' into pr/MrMDavidson/5949

jesserockz committed:
Fix always_inline

jesserockz committed:
Move define

jesserockz reviewed (changes requested):
I want to start by saying sorry it has taken so long to really make a start on reviewing this PR.

I have not reviewed this components actual code yet, I have just been fixing and updating it so CI will start passing etc. 

Can you split out the `display::Point`/`local coordinates` stuff into its own PR as there could be feedback specifically to that from others, and it affects all display components.

I think also splitting `text_align_to_string` to its own PR too will make that easier to merge.

jesserockz committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson commented:
> I want to start by saying sorry it has taken so long to really make a start on reviewing this PR.
>
> I have not reviewed this components actual code yet, I have just been fixing and updating it so CI will start passing etc.

Cheers, appreciate you getting the build back to green!
 
> Can you split out the `display::Point`/`local coordinates` stuff into its own PR as there could be feedback specifically to that from others, and it affects all display components.
> 
> I think also splitting `text_align_to_string` to its own PR too will make that easier to merge.

Will endeavour to get some time this weekend to split these out. Do you want me to split out the doc side of it, too?

jesserockz commented:
> Will endeavour to get some time this weekend to split these out. Do you want me to split out the doc side of it, too?

That definitely makes it easier to merge code and docs together

MrMDavidson commented:
> > Will endeavour to get some time this weekend to split these out. Do you want me to split out the doc side of it, too?
> 
> That definitely makes it easier to merge code and docs together

Done! See #7243, #7244 and esphome/esphome-docs#4123

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

MrMDavidson committed:
Merge branch 'dev' into add-graphical-layout-system

swoboda1337 commented:
Gone stale and do we really need this anymore after LVGL?
